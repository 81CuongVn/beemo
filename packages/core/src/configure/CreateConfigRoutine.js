/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import parseArgs from 'yargs-parser';
import { Routine } from 'boost';
import Driver from '../Driver';

import type { DriverContext } from '../types';

export default class CreateConfigRoutine extends Routine<Object, DriverContext> {
  driver: Driver;

  execute(): Promise<string> {
    const { name } = this.driver;

    this.task(`Loading external ${name} module config`, this.loadConfigFromFilesystem);
    this.task(`Loading local ${name} Beemo config`, this.extractConfigFromPackage);
    this.task(`Merging ${name} config objects`, this.mergeConfigs);
    this.task(`Creating temporary ${name} config file`, this.createConfigFile);

    return this.serializeTasks([]);
  }

  /**
   * Create a temporary configuration file or pass as an option.
   */
  createConfigFile(config: Object): Promise<string> {
    const { metadata } = this.driver;
    const configPath = path.join(this.context.root, metadata.configName);

    this.tool.debug(`Creating config file ${chalk.cyan(configPath)}`);

    this.context.configPaths.push(configPath);

    this.tool.emit('create-config-file', [configPath, config]);

    return fs.writeFile(configPath, this.driver.formatConfig(config)).then(() => configPath);
  }

  /**
   * Extract configuration from "beemo.<driver>" within the local project's package.json.
   */
  extractConfigFromPackage(configs: Object[]): Promise<Object[]> {
    const { name } = this.driver;
    const { config } = this.tool;

    this.tool.invariant(
      !!config[name],
      `Extracting ${chalk.magenta(name)} config from package.json "beemo" property`,
      'Exists, extracting',
      'Does not exist, skipping',
    );

    if (config[name]) {
      configs.push(config[name]);
    }

    this.tool.emit('load-package-config', [config]);

    return Promise.resolve(configs);
  }

  /**
   * Gather CLI arguments to pass to the configuration file.
   */
  getArgsToPass(): Object {
    this.tool.debug('Gathering arguments to pass to config file');

    return parseArgs(
      [...this.driver.options.args, ...this.context.args].map(value => String(value)),
    );
  }

  /**
   * Merge multiple configuration sources using the current driver.
   */
  mergeConfigs(configs: Object[]): Promise<Object> {
    this.tool.debug(
      `Merging ${chalk.magenta(this.driver.name)} config from ${configs.length} sources`,
    );

    const config = configs.reduce(
      (masterConfig, cfg) => this.driver.mergeConfig(masterConfig, cfg),
      {},
    );

    this.tool.emit('merge-config', [config]);

    return Promise.resolve(config);
  }

  /**
   * Load configuration from the node module (the consumer owned package).
   */
  loadConfigFromFilesystem(configs: Object[]): Promise<Object[]> {
    const { config: { module: moduleName }, configLoader } = this.tool;
    const { name } = this.driver;

    // Allow for local development
    const filePath =
      moduleName === '@local'
        ? path.join(this.context.root, `configs/${name}.js`)
        : configLoader.resolveModuleConfigPath(name, moduleName);
    const fileExists = fs.existsSync(filePath);

    this.tool.invariant(
      fileExists,
      `Loading ${chalk.magenta(name)} config from configuration module ${chalk.yellow(moduleName)}`,
      'Exists, loading',
      'Does not exist, skipping',
    );

    if (fileExists) {
      const config = configLoader.parseFile(filePath, this.getArgsToPass());

      this.tool.emit('load-module-config', [filePath, config]);

      configs.push(config);
    }

    return Promise.resolve(configs);
  }
}
