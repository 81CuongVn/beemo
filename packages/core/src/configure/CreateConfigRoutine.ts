/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import merge from 'lodash/merge';
import optimal, { instance, Struct } from 'optimal';
import { ConfigLoader, Routine } from '@boost/core';
import parseArgs from 'yargs-parser';
import Driver from '../Driver';
import DriverContext from '../contexts/DriverContext';
import { STRATEGY_COPY, STRATEGY_REFERENCE, STRATEGY_CREATE, STRATEGY_NATIVE } from '../constants';

export interface CreateConfigOptions {
  driver: Driver<any>;
}

export default class CreateConfigRoutine extends Routine<DriverContext, CreateConfigOptions> {
  bootstrap() {
    this.options = optimal(
      this.options,
      {
        driver: instance(Driver).required(),
      },
      {
        name: 'CreateConfigRoutine',
      },
    );
  }

  execute(): Promise<string> {
    const { metadata, name, options } = this.options.driver;
    const strategy =
      options.strategy === STRATEGY_NATIVE ? metadata.configStrategy : options.strategy;

    switch (strategy) {
      case STRATEGY_REFERENCE:
        this.task(`Referencing ${name} config file`, this.referenceConfigFile);
        break;

      case STRATEGY_COPY:
        this.task(`Copying ${name} config file`, this.copyConfigFile);
        break;

      case STRATEGY_CREATE:
        this.task(`Loading source ${name} module config`, this.loadConfigFromFilesystem);
        this.task(`Loading local ${name} Beemo config`, this.extractConfigFromPackage);
        this.task(`Merging ${name} config objects`, this.mergeConfigs);
        this.task(`Creating ${name} config file`, this.createConfigFile);
        break;

      default:
        this.skip(true);
        break;
    }

    return this.serializeTasks();
  }

  /**
   * Copy configuration file from module.
   */
  copyConfigFile(context: DriverContext): Promise<string> {
    const { metadata } = this.options.driver;
    const configLoader = new ConfigLoader(this.tool);
    const sourcePath = this.getSourceConfigPath(configLoader);
    const configPath = path.join(context.root, metadata.configName);

    if (!sourcePath) {
      throw new Error(
        'Cannot copy configuration file. Source file does not exist in configuration module.',
      );
    }

    const config = configLoader.parseFile(sourcePath);

    this.debug('Copying config file to %s', chalk.cyan(configPath));

    this.options.driver.config = config;

    context.configPaths.push(configPath);

    this.tool.emit('copy-config-file', [configPath, config]);

    return fs
      .copy(sourcePath, configPath, {
        overwrite: true,
      })
      .then(() => configPath);
  }

  /**
   * Create a temporary configuration file or pass as an option.
   */
  createConfigFile(context: DriverContext, config: object): Promise<string> {
    const { metadata } = this.options.driver;
    const configPath = path.join(context.root, metadata.configName);

    this.debug('Creating config file %s', chalk.cyan(configPath));

    this.options.driver.config = config as any;

    context.configPaths.push(configPath);

    this.tool.emit('create-config-file', [configPath, config]);

    return fs
      .writeFile(configPath, this.options.driver.formatConfig(config))
      .then(() => configPath);
  }

  /**
   * Extract configuration from "beemo.<driver>" within the local project's package.json.
   */
  extractConfigFromPackage(context: DriverContext, prevConfigs: Struct[]): Promise<Struct[]> {
    const { name } = this.options.driver;
    const { config } = this.tool;
    const configs = [...prevConfigs];

    this.debug.invariant(
      !!config[name],
      `Extracting ${chalk.magenta(name)} config from package.json "beemo" property`,
      'Exists, extracting',
      'Does not exist, skipping',
    );

    if (config[name]) {
      const pkgConfig = config[name];

      configs.push(pkgConfig);

      this.tool.emit('load-package-config', [pkgConfig]);
    }

    return Promise.resolve(configs);
  }

  /**
   * Return absolute file path for config file within configuration module,
   * or an empty string if it does not exist.
   */
  getSourceConfigPath(configLoader: ConfigLoader): string {
    const { root, workspaceRoot } = this.context;
    const {
      config: { module: moduleName },
    } = this.tool;
    const { name } = this.options.driver;

    // Allow for local development
    const filePath =
      moduleName === '@local'
        ? path.join(workspaceRoot || root, `configs/${name}.js`)
        : configLoader.resolveModuleConfigPath(name, moduleName);
    const fileExists = fs.existsSync(filePath);

    this.debug.invariant(
      fileExists,
      `Loading ${chalk.magenta(name)} config from configuration module ${chalk.yellow(moduleName)}`,
      'Exists, loading',
      'Does not exist, skipping',
    );

    return fileExists ? filePath : '';
  }

  /**
   * Merge multiple configuration sources using the current driver.
   */
  mergeConfigs(context: DriverContext, configs: Struct[]): Promise<Struct> {
    this.debug(
      'Merging %s config from %d sources',
      chalk.magenta(this.options.driver.name),
      configs.length,
    );

    const config = configs.reduce(
      (masterConfig, cfg) => this.options.driver.mergeConfig(masterConfig, cfg),
      {},
    );

    this.tool.emit('merge-config', [config]);

    return Promise.resolve(config);
  }

  /**
   * Load configuration from the node module (the consumer owned package).
   */
  loadConfigFromFilesystem(context: DriverContext): Promise<Struct[]> {
    const configLoader = new ConfigLoader(this.tool);
    const filePath = this.getSourceConfigPath(configLoader);
    const configs = [];

    if (filePath) {
      const args = merge({}, context.args, parseArgs(this.options.driver.getArgs()));
      const config = configLoader.parseFile(filePath, [args, this.tool]);

      this.tool.emit('load-module-config', [filePath, config]);

      configs.push(config);
    }

    return Promise.resolve(configs);
  }

  /**
   * Reference configuration file from module using a require statement.
   */
  referenceConfigFile(context: DriverContext): Promise<string> {
    const { metadata } = this.options.driver;
    const configLoader = new ConfigLoader(this.tool);
    const sourcePath = this.getSourceConfigPath(configLoader);
    const configPath = path.join(context.root, metadata.configName);

    if (!sourcePath) {
      throw new Error(
        'Cannot reference configuration file. Source file does not exist in configuration module.',
      );
    }

    const config = configLoader.parseFile(sourcePath);

    this.debug('Referencing config file to %s', chalk.cyan(configPath));

    this.options.driver.config = config;

    context.configPaths.push(configPath);

    this.tool.emit('reference-config-file', [configPath, config]);

    return fs
      .writeFile(
        configPath,
        `module.exports = require('./${path.relative(context.root, sourcePath)}');`,
      )
      .then(() => configPath);
  }
}
