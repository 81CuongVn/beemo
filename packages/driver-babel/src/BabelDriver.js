/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import path from 'path';
import rimraf from 'rimraf';
import { Driver } from '@beemo/core';

// Success: Writes file list to stdout
// Failure: Throws SyntaxError to stderr
export default class BabelDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'babel',
      configName: '.babelrc',
      description: 'Transpile files with Babel.',
      title: 'Babel',
    });

    this.setCommandOptions({
      clean: {
        boolean: true,
        default: true,
        description: 'Clean the target folder',
      },
    });

    this.on('babel.before-execute', this.handleCleanTarget);
  }

  /**
   * Automatically clean the target folder if --out-dir is used.
   */
  handleCleanTarget = (event: Event, driver: Driver, args: string[], context: Object) => {
    const options = context.yargs;

    if (options.clean && options.outDir) {
      rimraf.sync(path.resolve(options.outDir));
    }
  };
}
