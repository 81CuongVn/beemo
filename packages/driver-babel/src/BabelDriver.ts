/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import path from 'path';
import rimraf from 'rimraf';
import { Driver, DriverContext } from '@beemo/core';

// Success: Writes file list to stdout
// Failure: Throws SyntaxError to stderr
export default class BabelDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'babel',
      configName: '.babelrc',
      description: 'Transpile files with Babel.',
      title: 'Babel',
    });

    this.setCommandOptions({
      clean: {
        boolean: true,
        default: true,
        description: 'Clean the target folder',
      },
    });

    this.on('babel.before-execute', this.handleCleanTarget);
  }

  /**
   * Automatically clean the target folder if --out-dir is used.
   */
  handleCleanTarget = (driver: Driver, argList: string[], context: DriverContext) => {
    const { args } = context;

    if (args.clean && args.outDir) {
      rimraf.sync(path.resolve(args.outDir));
    }
  };
}
