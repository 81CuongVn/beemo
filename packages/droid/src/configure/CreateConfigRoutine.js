/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import chalk from 'chalk';
import fs from 'fs-extra';
import path from 'path';
import parseArgs from 'yargs-parser';
import { Routine } from 'boost';
import Engine from '../Engine';

import type { DroidContext } from '../types';

export default class CreateConfigRoutine extends Routine<Object, DroidContext> {
  engine: Engine;

  /**
   * Create a temporary configuration file or pass as an option.
   */
  createConfigFile(config: Object): Promise<string> {
    const { metadata } = this.engine;
    const configPath = path.join(this.context.root, metadata.configName);

    this.tool.debug(`Creating config file ${chalk.cyan(configPath)}`);

    this.context.configPaths.push(configPath);

    this.tool.emit('create-config', null, [configPath, config]);

    return fs.writeFile(configPath, this.engine.formatFile(config)).then(() => configPath);
  }

  /**
   * Create config by executing tasks in order.
   */
  execute(): Promise<string> {
    const { name } = this.engine;

    this.task(`Loading external ${name} module config`, this.loadConfigFromFilesystem);
    this.task(`Loading local ${name} Droid config`, this.extractConfigFromPackage);
    this.task(`Merging ${name} config objects`, this.mergeConfigs);
    this.task(`Creating temporary ${name} config file`, this.createConfigFile);

    return this.serializeTasks([]);
  }

  /**
   * Extract configuration from "droid.<engine>" within the local project's package.json.
   */
  extractConfigFromPackage(configs: Object[]): Promise<Object[]> {
    const { name } = this.engine;
    const { config } = this.tool;

    this.tool.invariant(
      !config[name],
      `Extracting ${chalk.magenta(name)} config from package.json "droid" block`,
      'Exists, extracting',
      'Does not exist, skipping',
    );

    if (config[name]) {
      configs.push(config[name]);
    }

    this.tool.emit('load-package-config', null, [config]);

    return Promise.resolve(configs);
  }

  /**
   * Gather CLI arguments to pass to the configuration file.
   */
  getArgsToPass(): Object {
    this.tool.debug('\tGathering arguments to pass to config file');

    const { args } = this.engine.options;

    return parseArgs([
      ...this.context.args,
      ...(Array.isArray(args) ? args : args.split(' ')),
    ].map(value => String(value)));
  }

  /**
   * Merge multiple configuration sources using the current engine.
   */
  mergeConfigs(configs: Object[]): Promise<Object> {
    this.tool.debug(`Merging config from ${configs.length} sources`);

    const config = configs.reduce((masterConfig, cfg) => (
      this.engine.mergeConfig(masterConfig, cfg)
    ), {});

    this.tool.emit('merge-config', null, [config]);

    return Promise.resolve(config);
  }

  /**
   * Load configuration from the node module (the consumer owned package).
   */
  loadConfigFromFilesystem(configs: Object[]): Promise<Object[]> {
    const { config: { config: moduleName }, configLoader } = this.tool;
    const { name } = this.engine;

    // Allow for local development
    const filePath = (moduleName === '@local')
      ? path.join(this.context.root, `config/${name}.js`)
      : configLoader.resolveModuleConfigPath(name, moduleName);
    const fileExists = fs.existsSync(filePath);

    this.tool.invariant(
      fileExists,
      `Loading config from configuration module ${chalk.yellow(moduleName)}`,
      'Exists, loading',
      'Does not exist, skipping',
    );

    if (fileExists) {
      const config = configLoader.parseFile(filePath, this.getArgsToPass());

      this.tool.debug(`\tParsing config from ${chalk.cyan(filePath)}`);

      this.tool.emit('load-module-config', null, [filePath, config]);

      configs.push(config);
    }

    return Promise.resolve(configs);
  }
}
