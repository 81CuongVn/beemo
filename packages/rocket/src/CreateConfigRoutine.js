/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Routine } from 'boost';

// TODO pass opts/args from command line / isolate scope
export default class CreateConfigRoutine extends Routine {
  bootstrap() {
    // TODO skip if file mtimes havent changed
    this
      .task('Loading module config', this.loadConfigFromFilesystem)
      .task('Loading project package.json config', this.extractConfigFromPackage)
      .task('Merging configs', this.mergeConfigs)
      .task('Creating config file', this.createConfigFile);
  }

  /**
   * Create a temporary configuration file to pass as a CLI argument to each engine's command.
   */
  createConfigFile(config: Config): string {
    // TODO create and return file path
  }

  /**
   * Create config by executing tasks in order.
   */
  execute(): Promise<*> /* TODO */ {
    return this.serializeTasks([]);
  }

  /**
   * Extract configuration from "rocket.<plugin>" within the local project's package.json.
   */
  extractConfigFromPackage(configs: Object[]): Object[] {
    const { config } = this.tool;
    const { engineName } = this.context;

    if (config[engineName]) {
      configs.push(config[engineName]);
    }

    return configs;
  }

  /**
   * Merge multiple configuration sources using the current engine.
   */
  mergeConfigs(configs: Object[]): Object {
    const { engine } = this.context;

    return configs.reduce((masterConfig, config) => engine.merge(masterConfig, config), {});
  }

  /**
   * Load configuration from the middle node module (the consumer owned package).
   */
  loadConfigFromFilesystem(configs: Object[]): Object[] {
    // TODO Use ConfigLoader???

    return configs;
  }
}
