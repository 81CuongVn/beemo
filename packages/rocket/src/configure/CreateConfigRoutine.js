/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import fs from 'fs-extra';
import path from 'path';
import parseArgs from 'yargs-parser';
import { Routine } from 'boost';
import Engine from '../Engine';

import type { RocketContext } from '../types';

export default class CreateConfigRoutine extends Routine<Object, RocketContext> {
  engine: Engine;

  /**
   * Create a temporary configuration file or pass as an option.
   */
  createConfigFile(config: Object): Promise<string> {
    const { metadata } = this.engine;
    const configPath = path.join(this.context.root, metadata.configName);

    this.context.configPaths.push(configPath);

    this.tool.emit('create-config', null, [configPath, config]);

    return fs.writeFile(configPath, this.engine.formatFile(config)).then(() => configPath);
  }

  /**
   * Create config by executing tasks in order.
   */
  execute(): Promise<string> {
    const { name } = this.engine;

    this.task(`Loading external ${name} module config`, this.loadConfigFromFilesystem);
    this.task(`Loading local ${name} rocket config`, this.extractConfigFromPackage);
    this.task(`Merging ${name} config objects`, this.mergeConfigs);
    this.task(`Creating temporary ${name} config file`, this.createConfigFile);

    return this.serializeTasks([]);
  }

  /**
   * Extract configuration from "rocket.<engine>" within the local project's package.json.
   */
  extractConfigFromPackage(configs: Object[]): Promise<Object[]> {
    const { name } = this.engine;
    const { config } = this.tool;

    if (config[name]) {
      configs.push(config[name]);
    }

    this.tool.emit('load-package-config', null, [config]);

    return Promise.resolve(configs);
  }

  /**
   * Gather CLI arguments to pass to the configuration file.
   */
  getArgsToPass(): Object {
    return parseArgs([
      ...this.context.args,
      ...this.engine.options.args,
    ].map(value => String(value)));
  }

  /**
   * Merge multiple configuration sources using the current engine.
   */
  mergeConfigs(configs: Object[]): Promise<Object> {
    const config = Promise.resolve(configs.reduce((masterConfig, config) => (
      this.engine.mergeConfig(masterConfig, config)
    ), {}));

    this.tool.emit('merge-config', null, [config]);

    return config;
  }

  /**
   * Load configuration from the node module (the consumer owned package).
   */
  loadConfigFromFilesystem(configs: Object[]): Promise<Object[]> {
    const { config: { config: moduleName }, configLoader } = this.tool;
    const { name } = this.engine;

    // Allow for local development
    const filePath = (moduleName === '@local')
      ? path.join(this.context.root, `config/${name}.js`)
      : configLoader.resolveModuleConfigPath(name, moduleName);

    if (fs.existsSync(filePath)) {
      const config = configLoader.parseFile(filePath, this.getArgsToPass());

      this.tool.emit('load-module-config', null, [filePath, config]);

      configs.push(config);
    }

    return Promise.resolve(configs);
  }
}
