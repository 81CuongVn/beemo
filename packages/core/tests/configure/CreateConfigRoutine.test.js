import fs from 'fs-extra';
import CreateConfigRoutine from '../../src/configure/CreateConfigRoutine';
import BabelDriver from '../../../driver-babel/src/BabelDriver';

jest.mock('fs-extra');

describe('CreateConfigRoutine', () => {
  let routine;
  let driver;

  beforeEach(() => {
    driver = new BabelDriver({ args: ['--bar'] });
    driver.name = 'babel';

    routine = new CreateConfigRoutine('babel', 'Configure Babel', { driver });
    routine.context = {
      args: ['--foo'],
      configPaths: [],
      root: '/',
    };
    routine.tool = {
      config: {
        module: '@local',
      },
      debug() {},
      emit() {},
      invariant() {},
      on() {},
    };

    driver.tool = routine.tool;
    driver.bootstrap();
  });

  beforeEach(() => {
    fs.existsSync.mockImplementation(() => true);
    fs.writeFile.mockImplementation(() => Promise.resolve());
  });

  describe('execute()', () => {
    it('executes pipeline and returns final config path', async () => {
      // Step 1
      routine.tool.configLoader = {
        parseFile: () => ({ foo: 123 }),
      };

      // Step 2
      routine.tool.config.babel = { bar: 'abc' };

      const path = await routine.execute();

      expect(path).toBe('/.babelrc');
      expect(fs.writeFile).toHaveBeenCalledWith(
        '/.babelrc',
        JSON.stringify({ foo: 123, bar: 'abc' }, null, 2),
      );
    });
  });

  describe('bootstrap()', () => {
    it('errors if no driver defined', () => {
      delete routine.config.driver;

      expect(() => {
        routine.bootstrap();
      }).toThrowError(
        'Invalid CreateConfigRoutine option "driver". Field is required and must be defined.',
      );
    });
  });

  describe('createConfigFile()', () => {
    it('adds path to context', async () => {
      await routine.createConfigFile({ foo: 'bar' });

      expect(routine.context.configPaths).toEqual(['/.babelrc']);
    });

    it('writes and formats file', async () => {
      const path = await routine.createConfigFile({ foo: 'bar' });

      expect(fs.writeFile).toHaveBeenCalledWith('/.babelrc', '{\n  "foo": "bar"\n}');
      expect(path).toBe('/.babelrc');
    });

    it('triggers `create-config-file` event', async () => {
      const spy = jest.spyOn(routine.tool, 'emit');

      await routine.createConfigFile({ foo: 'bar' });

      expect(spy).toHaveBeenCalledWith('create-config-file', ['/.babelrc', { foo: 'bar' }]);
    });
  });

  describe('extractConfigFromPackage()', () => {
    it('does nothing if config does not exist', async () => {
      const configs = await routine.extractConfigFromPackage([]);

      expect(configs).toEqual([]);
    });

    it('extracts config by name', async () => {
      routine.tool.config.babel = { foo: 'bar' };

      const configs = await routine.extractConfigFromPackage([]);

      expect(configs).toEqual([{ foo: 'bar' }]);
    });

    it('triggers `load-package-config` event', async () => {
      const spy = jest.spyOn(routine.tool, 'emit');

      routine.tool.config.babel = { foo: 'bar' };

      await routine.extractConfigFromPackage([]);

      expect(spy).toHaveBeenCalledWith('load-package-config', [{ foo: 'bar' }]);
    });

    it('doesnt trigger `load-package-config` if no config', async () => {
      const spy = jest.spyOn(routine.tool, 'emit');

      await routine.extractConfigFromPackage([]);

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('mergeConfigs()', () => {
    it('merges multiple sources', async () => {
      const config = await routine.mergeConfigs([
        { foo: 123, qux: true },
        { bar: 'abc' },
        { foo: 456 },
      ]);

      expect(config).toEqual({
        foo: 456,
        bar: 'abc',
        qux: true,
      });
    });

    it('calls `mergeConfig` on driver', async () => {
      const spy = jest.spyOn(driver, 'mergeConfig');

      await routine.mergeConfigs([{ foo: 123, qux: true }, { bar: 'abc' }, { foo: 456 }]);

      expect(spy).toHaveBeenCalledTimes(3);
    });

    it('triggers `merge-config` event with final config object', async () => {
      const spy = jest.spyOn(routine.tool, 'emit');

      const config = await routine.mergeConfigs([
        { foo: 123, qux: true },
        { bar: 'abc' },
        { foo: 456 },
      ]);

      expect(spy).toHaveBeenCalledWith('merge-config', [config]);
    });
  });

  describe('getArgsToPass()', () => {
    it('merges driver and context args', () => {
      expect(routine.getArgsToPass()).toEqual({
        _: [],
        foo: true,
        bar: true,
      });
    });
  });

  describe('loadConfigFromFilesystem()', () => {
    beforeEach(() => {
      routine.tool.configLoader = {
        resolveModuleConfigPath: jest.fn(
          (name, moduleName) => `/node_modules/${moduleName}/configs/${name}.js`,
        ),
        parseFile: jest.fn(filePath => ({ filePath })),
      };
    });

    it('loads config if it exists', async () => {
      const configs = await routine.loadConfigFromFilesystem([]);

      expect(configs).toEqual([{ filePath: '/configs/babel.js' }]);
    });

    it('does nothing if config does not exist', async () => {
      fs.existsSync.mockImplementation(() => false);

      const configs = await routine.loadConfigFromFilesystem([]);

      expect(configs).toEqual([]);
    });

    it('uses local path when using @local config', async () => {
      const configs = await routine.loadConfigFromFilesystem([]);

      expect(configs).toEqual([{ filePath: '/configs/babel.js' }]);
    });

    it('uses module path when using custom config', async () => {
      routine.tool.config.module = 'foo-bar';

      const configs = await routine.loadConfigFromFilesystem([]);

      expect(configs).toEqual([{ filePath: '/node_modules/foo-bar/configs/babel.js' }]);
    });

    it('triggers `load-module-config` event', async () => {
      const spy = jest.spyOn(routine.tool, 'emit');

      await routine.loadConfigFromFilesystem([]);

      expect(spy).toHaveBeenCalledWith('load-module-config', [
        '/configs/babel.js',
        { filePath: '/configs/babel.js' },
      ]);
    });

    it('doesnt trigger `load-module-config` event if files does not exist', async () => {
      fs.existsSync.mockImplementation(() => false);

      const spy = jest.spyOn(routine.tool, 'emit');

      await routine.loadConfigFromFilesystem([]);

      expect(spy).not.toHaveBeenCalled();
    });

    it('parses file with yargs options', async () => {
      await routine.loadConfigFromFilesystem([]);

      expect(routine.tool.configLoader.parseFile).toHaveBeenCalledWith('/configs/babel.js', {
        _: [],
        foo: true,
        bar: true,
      });
    });
  });
});
