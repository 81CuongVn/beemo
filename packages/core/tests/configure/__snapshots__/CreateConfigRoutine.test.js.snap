// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateConfigRoutine bootstrap() errors if no driver defined 1`] = `"Invalid CreateConfigRoutine field \\"driver\\". Field is required and must be defined."`;

exports[`CreateConfigRoutine copyConfigFile() errors if no source file 1`] = `"Cannot copy configuration file. Source file does not exist in configuration module."`;
